### PR TITLE
Exec input with Graphite protocol support

### PR DESCRIPTION
**Pull request by henrypfhu**
According to @sparrc suggestion, I combined the graphite feature into the exec plugin, and refactor the parser module into internal/encoding directory.

I open this new PR for exec plugin. 

Old PR please refer to #627

### CODE REVIEW

File: plugins/inputs/exec/README.md [hunk 1]
```diff
@@ -11,21 +27,44 @@ and strings will be ignored.
 ```
 # Read flattened metrics from one or more commands that output JSON to stdout
 [[inputs.exec]]
-  # the command to run
-  command = "/usr/bin/mycollector --foo=bar"
+  # Shell/commands array
+  commands = ["/tmp/test.sh","/tmp/test2.sh"]
```
Comment (sparrc) [inline plugins/inputs/exec/README.md]: This is a breaking change. You'll need to change unit tests for this. Also `exec` should continue to support users that have the `command = ` specified for backwards compatibility.

File: internal/encoding/graphite/innerconfig.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+)
+
+const (
+	// DefaultSeparator is the default join character to use when joining multiple
+	// measurment parts in a template.
+	DefaultSeparator = "."
+)
+
+// Config represents the configuration for Graphite endpoints.
+type InnerConfig struct {
```
Comment (sparrc) [inline internal/encoding/graphite/innerconfig.go]: call this just `config`, not `innerconfig`

File: plugins/inputs/exec/exec.go [hunk 3]
```diff
@@ -60,47 +103,86 @@ func NewExec() *Exec {
...
+	return nil
+}
+
 func (e *Exec) SampleConfig() string {
 	return sampleConfig
 }
 
 func (e *Exec) Description() string {
-	return "Read flattened metrics from one or more commands that output JSON to stdout"
+	return "Read metrics from one or more commands that output graphite line protocol to stdout"
```
Comment (sparrc) [inline plugins/inputs/exec/exec.go]: also mention JSON, influx

File: internal/encoding/graphite/config.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+const (
+	// DefaultSeparator is the default join character to use when joining multiple
+	// measurment parts in a template.
+	DefaultSeparator = "."
+)
+
+// Config represents the configuration for Graphite endpoints.
+type Config struct {
+	Separator string
+	Tags      []string
```
Comment (sparrc) [inline internal/encoding/graphite/config.go]: Please remove the `Tags` capability from the graphite parser. Telegraf already has the capability to specify per-plugin tags: https://github.com/influxdata/telegraf/blob/master/CONFIGURATION.md#input-config-tags

File: internal/encoding/parser.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+type Parser struct {
+	graphiteParser *graphite.Parser
+}
+
+func NewParser(parser *graphite.Parser) *Parser {
+	return &Parser{graphiteParser: parser}
+}
+
+func (p *Parser) Parse(dataFormat string, out []byte, acc telegraf.Accumulator) error {
```
Comment (sparrc) [inline internal/encoding/parser.go]: `Parse` functions should _not_ be specific to any plugin, just make two functions here with signatures:

```
func (p *Parser) Parse(dataFormat string, buf []byte) ([]telegraf.Metric, error)
func (p *Parser) ParseLine(dataFormat, line string) (telegraf.Metric, error)
```

Comment (sparrc) [inline internal/encoding/parser.go]: and they should both support JSON, influx, and graphite

File: internal/encoding/parser.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+	var err error
+	var metrics []telegraf.Metric
+	var metric telegraf.Metric
+
+	switch dataFormat {
+	case "", "json":
```
Comment (sparrc) [inline internal/encoding/parser.go]: Make this case only `case "json":`, if plugins want to convert "" -> "json" they can make that assumption themselves.

**Comment by sparrc** [comment]
@henrypfhu This is looking very good so far, I like the structure I just have a few comments on the generic parser.

Since the graphite template parser was mostly taken from the influxdb repo, can you also add the unit tests from that repo? (you might need to remove a few of them because of the tags removal)

File: README.md [hunk 7]
```diff
@@ -158,7 +158,9 @@ Currently implemented sources:
 * disque
 * docker
 * elasticsearch
-* exec (generic JSON-emitting executable plugin)
+* exec (generic line-protocol-emitting executable plugin, support JSON, influx and graphite)
+* socket (generic line protocol listen input service, support influx and graphite)
```
Comment (sparrc) [inline README.md]: this PR doesn't include `socket` or `tail`

**Comment by henrypfhu** [comment]
Hi @sparrc, from tomorrow I will be in vacation fro Chinese Spring Festival, these days I will be offline. Unit test cases will be postponed. Actually, I already tested all scenarios manually. So far I fixed all bugs I could found. Thx...

**Comment by sparrc** [comment]
Thanks for the heads-up on your holiday time @henrypfhu, hope you enjoy it :-)

I squashed your commits into one and am now working on a new branch based off your changes (see #655). I'm going to close this PR because I have some ideas and changes that I would like to make on how the generic parsers feature is going to look, and it's also very relevant to an MQTT consumer plugin that I was coincidentally currently working on.

Thanks for your great work getting this started!

**Pull request closed without merging**